### PR TITLE
loader: support concurrent restore at table level

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
@tiancaiamao @qiuyesuifeng @andelf PTAL

### CODE REVIEW

File: loader/loader.go [hunk 1]
```diff
@@ -314,172 +313,145 @@ func (l *Loader) dispatchSQL(file string, schema string, table string, checkExis
...
-				}
-
-				count = 0
-				sqls = sqls[0:0]
-				lastSyncTime = now
-
-				skipConstraintCheck = true
+			sqls := make([]string, 0, 2)
+			sqls = append(sqls, fmt.Sprintf("USE %s;", job.schema))
+			sqls = append(sqls, job.sql)
```
Comment (nolouch) [inline loader/loader.go]: Why remove batch? maybe allocate a lot of memory and generate a large transaction when the table is very large.

File: loader/loader.go [hunk 2]
```diff
@@ -314,172 +313,145 @@ func (l *Loader) dispatchSQL(file string, schema string, table string, checkExis
...
+			if err := executeSQL(conn, sqls, true, job.skipConstraintCheck); err != nil {
```
Comment (GregoryIan) [inline loader/loader.go]: I feel `session.tidb_skip_constraint_check` has some problem, how to use it? while u set it, It still return an error and we can't get an accurate and detail description about it

File: loader/loader.go (hunk 1 above)
Comment (AndreMouche) [inline loader/loader.go]: `job.sql` is already a batch with size=1M @nolouch

File: loader/loader.go [hunk 3]
```diff
@@ -314,172 +313,145 @@ func (l *Loader) dispatchSQL(file string, schema string, table string, checkExis
...
+ *                        main-thread
+ * |        table-worker       |         table-worker       |
+ * | data-worker | data-worker |  data-worker | data-worker |
+ */
 
-	skipConstraintCheck := true
+func (l *Loader) runTableWorker(poolId int, queue chan *tableJob) {
 	for {
 		select {
 		case job, ok := <-queue:
```
Comment (tiancaiamao) [inline loader/loader.go]: No need select here, it can be write directly:
```
job, ok := <-queue:
```

File: loader/loader.go [hunk 4]
```diff
@@ -314,172 +313,145 @@ func (l *Loader) dispatchSQL(file string, schema string, table string, checkExis
...
-				lastSyncTime = time.Now()
+			l.tableJobWg.Done()
+		}
+	}
+}
 
-				skipConstraintCheck = true
+func (l *Loader) runDataWorker(poolId int, conn *Conn, queue chan *dataJob) {
+	for {
+		select {
```
Comment (tiancaiamao) [inline loader/loader.go]: ditto

File: loader/loader.go (hunk 1 above)
Comment (zhangjinpeng1987) [inline loader/loader.go]: To make the code simple.

File: loader/config.go [hunk 5]
```diff
@@ -30,7 +30,8 @@ func NewConfig() *Config {
 	fs.StringVar(&cfg.Dir, "d", "./", "Directory of the dump to import")
 
 	fs.IntVar(&cfg.Batch, "q", 1, "Number of queries per transaction, default 1")
```
Comment (nolouch) [inline loader/config.go]: remove this

**Comment by zhangjinpeng1987** [comment]
We also support split a big table into multiple blocks and restore concurrently.

File: loader/checkpoint.go [hunk 6]
```diff
@@ -23,31 +23,81 @@ import (
 
 	"github.com/juju/errors"
 	"github.com/ngaut/log"
+	"strconv"
```
Comment (tiancaiamao) [inline loader/checkpoint.go]: move to line 23, keep the format uniform.

File: loader/checkpoint.go [hunk 7]
```diff
@@ -23,31 +23,81 @@ import (
...
 )
 
 // CheckPoint represents checkpoint status
 type CheckPoint struct {
 	sync.RWMutex
-	path              string
-	restoredFiles     map[string]struct{}
-	restoreFromLastCP bool // restore from last checkpoint
+	path            string
+	restoredFiles   map[string]map[string]Set
```
Comment (tiancaiamao) [inline loader/checkpoint.go]: what about encapsulation the struct and provide method to operate those struct?

File: loader/loader.go [hunk 8]
```diff
@@ -21,71 +21,216 @@ import (
...
 
-	conns []*Conn
+				p.wg.Done()
+			}
+		}(i)
+	}
 
-	jobsQueue chan *job
+	// pool main routine
+	go func() {
```
Comment (tiancaiamao) [inline loader/loader.go]: remove this line and let the caller do `go workerPool.run()` ?

File: loader/loader.go [hunk 9]
```diff
@@ -97,13 +242,40 @@ func (l *Loader) Restore() error {
 		return errors.Trace(err)
 	}
 
+	// check last file-num-per-block and current file-num-per-block, they must equal
+	if l.checkPoint.FileNumPerBlock < 0 {
```
Comment (tiancaiamao) [inline loader/loader.go]: <= 0

**Comment by tiancaiamao** [comment]
LGTM @nolouch @AndreMouche

File: loader/checkpoint.go (hunk 7 above)
Comment (zhangjinpeng1987) [inline loader/checkpoint.go]: no need make it more complicated.

**Comment by zhangjinpeng1987** [comment]
@nolouch @AndreMouche PTAL

**Comment by zhangjinpeng1987** [comment]
@iamxy PTAL

**Comment by zhangjinpeng1987** [comment]
@nolouch @iamxy @andelf PTAL

**Comment by nolouch** [comment]
LGTM

**Pull request merged**